src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,276 @@
+        /// for Spark versions 2.3.0 and 2.3.1 and returns the JVMObjectReference object.
+        /// </summary>
+        /// <param name="javaSparkContext">Java Spark context object</param>
+        /// <param name="tempDir">Path where file is to be created</param>
+        /// <param name="value">Broadcast value of type object</param>
+        /// <returns></returns>
+        private JvmObjectReference CreateBroadcast_V2_3_1_Below(
+            JvmObjectReference javaSparkContext,
+            string tempDir,
+            object value)
+        {
+            WriteBroadcastValueToFile(tempDir, value);
+            
+            return (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "readBroadcastFromFile",
+                javaSparkContext,
+                _path);
+        }
+
+        /// <summary>
+        /// Calls the necessary Spark functions to create org.apache.spark.broadcast.Broadcast
+        /// object for Spark versions 2.3.2 and above, and returns the JVMObjectReference object.
+        /// </summary>
+        /// <param name="javaSparkContext">Java Spark context object</param>
+        /// <param name="sc">SparkContext object</param>
+        /// <param name="tempDir">Path where file is to be created</param>
+        /// <param name="value">Broadcast value of type object</param>
+        /// <returns></returns>
+        private JvmObjectReference CreateBroadcast_V2_3_2_Above(
+            JvmObjectReference javaSparkContext,
+            SparkContext sc,
+            string tempDir,
+            object value)
+        {
+            var pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+            
+            SparkConf sparkConf = sc.GetConf();
+            bool encryptionEnabled = bool.Parse(
+                sparkConf.Get("spark.io.encryption.enabled", "false"));
+
+            if (encryptionEnabled)
+            {
+                throw new NotImplementedException("Broadcast encryption is not supported yet.");
+            }
+            else
+            {
+                WriteBroadcastValueToFile(tempDir, value);
+            }
+
+            return (JvmObjectReference)javaSparkContext.Jvm.CallNonStaticJavaMethod(
+                javaSparkContext,
+                "broadcast",
+                pythonBroadcast);
+        }
+    }
+
+    /// <summary>
+    /// Global registry to store the object value of all active broadcast variables from
+    /// the workers. This registry is only used on the worker side when Broadcast.Value() is called
+    /// through a UDF.
+    /// </summary>
+    internal static class BroadcastRegistry
+    {
+        private static readonly ConcurrentDictionary<long, object> s_registry =
```

Comment:
Should this also be `ThreadStatic`?

Reply:
It is not in pyspark, so I don't think so. We want this to be consistent across the worker regardless of threads, I believe.